teps/0091-trusted-resources.md

```diff
@@ -230,71 +215,168 @@ spec:
+}
+
+type VerificationPolicySpec struct {
+	// Resources defines the patterns of Resources names that should be subject to this policy. For example, we may want to apply this Policy only from a certain github repo. Then the ResourcesPattern should include the path.
+	Resources []ResourcesPattern `json:"resources"`
+	// Authorities defines the rules for validating signatures.
+	Authorities []Authority `json:"authorities"`
+}
+
+type ResourcesPattern struct {
+	// Pattern defines a resource pattern.
+	Pattern string `json:"glob"`
+}
+
+// The authorities block defines the rules for discovering and
+// validating signatures.
+type Authority struct {
+	// Name is the name for this authority.
+	Name string `json:"name"`
+	// Key defines the type of key to validate the resource.
+	// +optional
+	Key *KeyRef `json:"key,omitempty"`
+	// Keyless sets the configuration to verify the authority against a Fulcio instance.
+	// +optional
+	Keyless *KeylessRef `json:"keyless,omitempty"`
+	// Sources sets the configuration to specify the sources from where to consume the signatures if the signature is not stored in the resource.
+	// +optional
+	Sources []Source `json:"source,omitempty"`
+}
+
+type KeyRef struct {
+	// SecretRef sets a reference to a secret with the key.
+	// +optional
+	SecretRef *v1.SecretReference `json:"secretRef,omitempty"`
+	// Data contains the inline public key.
+	// +optional
+	Data string `json:"data,omitempty"`
+	// KMS contains the KMS url of the public key
+	// Supported formats differ based on the KMS system used.
+	// +optional
+	KMS string `json:"kms,omitempty"`
+}
+
+type KeylessRef struct {
+	// URL defines a url to the keyless instance.
+	// +optional
+	URL *apis.URL `json:"url,omitempty"`
+	// Identities sets a list of identities.
+	// +optional
+	Identities []Identity `json:"identities,omitempty"`
+	// CACert sets a reference to CA certificate
+	// +optional
+	CACert *KeyRef `json:"ca-cert,omitempty"`
+}
+```
+
 
-### Webhook and Configuration
+### Configuration
 
-Validating webhook can use configmap to allow users to config when the Resource fails the verification
-1) Directly fail the run;
-2) Not fail the run, and when Tekton Pipeline's dependent knative version support admission webhook warnings, return warning in `apis.FieldError`.
-3) Skip the validation;
+Reconciler will respond to verification failure based on configMap settings to:
+1) Return error in reconciler, mark the status as error with `FailedVerification` reason;
+2) Not return error, but mark the TaskRun/PipelineRun as `FailedVerification`;
+3) Skip the verification;
 
 Configuration sample:
 ```yaml
 apiVersion: v1
 kind: ConfigMap
 metadata:
-  name: config-trusted-resources
+  name: feature-flag
   namespace: tekton-pipelines
 data:
-  onError: "stopAndFail"
-  cosign-pubkey-path: "/etc/signing-secrets/cosign.pub"
-  kms-pubkey-path: "gcpkms://projects/<project>/locations/<location>/keyRings/<keyring>/cryptoKeys/<key>"
+  resources-verify: true
```

Comment:
I don't think you need feature flag, assuming that the default behavior of the cluster remains the same if there is no policy defined.

Creating the policy is effectively enabling the feature.

Reply:
oh, I was assuming they wouldn't be marked as generic invalid - it'd be great to have a different error state with an actionable message like `VerificationFailed: denied by policy default/verification-policy` so it's not a complete surprise why a Run is newly failing.

Because adding the policy config is already an opt-in action, also requiring a feature flag seemed redundant to me. But if you want to include it that's fine too!